src/slice.c

```diff
@@ -207,34 +207,39 @@ r_obj* df_slice(r_obj* x, r_obj* subscript) {
   return out;
 }
 
+static
+r_obj* bracket_dispatch(r_obj* x, r_obj* subscript) {
+  return vctrs_dispatch2(
+    syms_bracket, fns_bracket,
+    syms_x, x,
+    syms_i, subscript
+  );
+}
+static
+r_obj* bracket_shaped_dispatch(r_obj* x, r_obj* subscript) {
+  return vctrs_dispatch2(
+    syms.bracket_shaped_dispatch, fns.bracket_shaped_dispatch,
+    syms_x, x,
+    syms_i, subscript
+  );
+}
```

Comment:
We now have these nice two parallel functions. `bracket_dispatch()` calls `x[i]` and `bracket_shaped_dispatch()` just calls `x[i, , , drop = FALSE]` with as many `,` as needed.